db-engine/src/main/java/com/splicemachine/db/iapi/types/ListDataType.java

```diff
@@ -0,0 +1,565 @@
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow;
+import org.apache.spark.sql.catalyst.expressions.codegen.UnsafeArrayWriter;
+import org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.StructField;
+
+import java.io.*;
+import java.sql.ResultSet;
+
+/**
+ * ListDataType satisfies the DataValueDescriptor
+ * interfaces (i.e., OrderableDataType). It implements a list of
+ * DataValueDescriptors, for example (SQLLongint, SQLTimestamp, SQLChar).
+ * <p>
+ * The main purpose for this class is to facilitate representation
+ * of multicolumn IN lists, e.g. (col_1, col2) IN ((1,2), (3,4))
+ * One use of this is the conversion of equality predicates in DNF form,
+ * e.g.  (c1 = 1 and c2 = 2) or (c1 = 2 and c2 = 3)
+ * ... into an IN list predicate:  (c1,c2) in ((1,2), (2,3))
+ * in order to enable optimizations such as MultiProbeTableScan.
+ * <p>
+ * A comparison operation between two ListDataTypes is true iff
+ * the two ListDataTypes being compared have the same number of
+ * elements and the comparison operation is true for each pairwise
+ * comparison of same-positioned elements from the two lists,
+ * for example, the comparison:
+ * (1, '2001-01-01', 'Widget') = (1, '2001-01-01', 'Sprocket')
+ * is evaluated as the following boolean expression:
+ * (1 = 1) AND ('2001-01-01' = '2001-01-01') and 'Widget' = 'Sprocket'
+ * <p>
+ * If any of the elements of the ListDataType holds a null, the
+ * ListDataType item itself can be considered as a null value since
+ * (null OP X) AND (A OP B) and (C OP D) ... is never TRUE,
+ * and a multicolumn IN list is never negated (we never apply
+ * the NOT operator on it).
+ * <p>
+ */
+public final class ListDataType extends DataType {
+
+	int numElements = 0;
+	DataValueDescriptor[] dvd = null;
+
+    @Override
+	public int getLength()
+	{
+		return numElements;
+	}
+
+	public void setLength(int len) {
+        numElements = len;
+        dvd = new DataValueDescriptor[len];
+    }
+
+
+    // this is for DataType's error generator
+    @Override
+    public String getTypeName() {
+        return TypeId.LIST_NAME;
+    }
+
+    @Override
+    public String getString() throws StandardException
+    {
+        String theString = new String();
+        theString.concat("( ");
+        for (int i = 0; i < numElements; i++) {
+            String stringToAdd = dvd[i].getString();
+            theString.concat(stringToAdd != null ? stringToAdd : "NULL");
+            if (i != numElements - 1)
+                theString.concat(", ");
+        }
+        theString.concat(")");
+        return theString;
+    }
+
+    @Override
+    public String getTraceString() throws StandardException
+    {
+        return getString();
+    }
+
+    @Override
+    public boolean isNull() {
+        boolean localIsNull = true;
+        
+        if (numElements > 1) {
```

Comment:
Why the check of numElements > 1 here? if numElements=1, we may not set the nullability correctly.

Reply:
Removed the if condition.